src/Cli/dotnet/Program.cs

```diff
@@ -204,34 +225,52 @@ internal static int ProcessArgs(string[] args, ITelemetry telemetryClient = null
                 Console.WriteLine($"Telemetry is: {(telemetryClient.Enabled ? "Enabled" : "Disabled")}");
             }
 
+            PerformanceLogEventSource.Log.TelemetrySendIfEnabledStart();
             TelemetryEventEntry.SendFiltered(topLevelCommandParserResult);
+            PerformanceLogEventSource.Log.TelemetrySendIfEnabledStop();
 
             int exitCode;
             if (BuiltInCommandsCatalog.Commands.TryGetValue(topLevelCommandParserResult.Command, out var builtIn))
             {
+                PerformanceLogEventSource.Log.BuiltInCommandParserStart();
                 var parseResult = Parser.Instance.ParseFrom($"dotnet {topLevelCommandParserResult.Command}", appArgs.ToArray());
+                PerformanceLogEventSource.Log.BuiltInCommandParserStop();
+
                 if (!parseResult.Errors.Any())
                 {
+                    PerformanceLogEventSource.Log.TelemetrySendIfEnabledStart();
                     TelemetryEventEntry.SendFiltered(parseResult);
+                    PerformanceLogEventSource.Log.TelemetrySendIfEnabledStop();
                 }
 
+                PerformanceLogEventSource.Log.BuiltInCommandStart();
```

Comment:
you need to create a new method for every new kind of log?

Reply:
Right.  The EventSource pattern is that you create a method on the EventSource for each event that you want to emit.  During initialization, EventSource will iterate over all of these methods via reflection, and build the metadata necessary to emit the events.  Then, emission of the events themselves is cheap.

---
In reply to: [505726520](https://github.com/wli3/sdk/pull/1#discussion_r505726520) [](ancestors = 505726520)

Reply:
Could you elaborate the benefit of using EventSource class for this specific feature? Since it also does not make talking to ETW easier. From the docs, looks like it is about strong typing telemetry. But I don't think it is too important for this feature. The existing pipeline uses [events as well](https://github.com/dotnet/sdk/blob/b1223209644d900702287faea8e9b71f95ec49f8/src/Cli/Microsoft.DotNet.Cli.Utils/TelemetryEventEntry.cs).

I've worked on web services before. And the existing pipeline could solve many things I saw in production.

1. Test telemetry
2. Do not pollute business logic  (so does EventSource)
3. Enrich telemetry in a different file with desire format
4. Less coupling, via event, looks like EventSource does it too.

Reply:
I think I may have misunderstood your question about ETW.  EventSource does emit these events to ETW - essentially, EventSource is the built-in mechanism that will do so.  It will also emit the events to other performance-related event sinks like EventPipe and LTTng.  There just isn't any ETW specific code that I had to implement as part of this PR.